Mathlib/Analysis/Calculus/FirstDerivativeTest.lean

```diff
@@ -0,0 +1,237 @@
+    apply ContinuousAt.continuousWithinAt
+    exact DifferentiableAt.continuousAt hD
+
+/-- If `f` is continuous at `b` and differentiable on `(b,c)` then `f` is
+  continuous on the half-open interval `[b,c)`. -/
+theorem continuous_Ico.{u_1, u_2, u_3} {𝕜 : Type u_1} [NontriviallyNormedField 𝕜]
+  {E : Type u_2} [NormedAddCommGroup E] [NormedSpace 𝕜 E]
+    [LinearOrder E] [OrderClosedTopology E]
+    {F : Type u_3} [NormedAddCommGroup F] [NormedSpace 𝕜 F]
+    {f : E → F}
+    {b c : E} (g₁ : b < c)
+    (h : ContinuousAt f b) (hd₁ : DifferentiableOn 𝕜 f (Set.Ioo b c)) :
+    ContinuousOn f (Set.Ico b c) := by
+  intro x hx
+  by_cases H : x = b
+  · subst H;simp_all;exact ContinuousAt.continuousWithinAt h
+  · have hab: x ∈ Set.Ioo b c := by
+      refine Set.mem_Ioo.mpr ?_;
+      constructor
+      · contrapose H; simp_all;
+        apply le_antisymm;tauto;tauto
+      · exact hx.2
+    have hD : DifferentiableAt 𝕜 f x := by
+      apply differentiableOn_differentiableAt_Ioo;repeat tauto
+    apply ContinuousAt.continuousWithinAt
+    · exact DifferentiableAt.continuousAt hD
+
+/-- If `f` is differentiable on a set `s` then so is `-f`. -/
+theorem differentiableOn_neg_Ioo
+  {f : ℝ → ℝ} {s : Set ℝ} (hd₀ : DifferentiableOn ℝ f s) :
+    DifferentiableOn ℝ (-f) s := by
+  show DifferentiableOn ℝ ((fun x => -x) ∘ (fun x => f x)) s
+  apply DifferentiableOn.comp
+  · apply differentiableOn_neg
+  · tauto
+  · show Set.MapsTo f s Set.univ
+    exact fun ⦃x⦄ _ ↦ trivial
+
+/-- If `f'` is the derivative of `f` then  `f' x ≤ 0 → 0 ≤ (-f)' x`. -/
+theorem deriv_neg_nonneg {f : ℝ → ℝ} {a b : ℝ}
+  (hd₀ : DifferentiableOn ℝ f (Set.Ioo a b))
+    (h₀ : ∀ x ∈ Set.Ioo a b, deriv f x ≤ 0) (x : ℝ) (hx : x ∈ Set.Ioo a b)
+    : 0 ≤ deriv (-f) x := by
+  show 0 ≤ deriv (((fun x => -x) ∘ (fun x => f x))) x
+  rw [deriv.comp]
+  simp
+  apply h₀
+  tauto
+  refine Differentiable.differentiableAt ?hh₂.h
+  · exact differentiable_neg
+  · apply DifferentiableOn.differentiableAt
+    · exact hd₀
+    · refine Ioo_mem_nhds ?hh.hs.ha ?hh.hs.hb
+      · exact hx.1
+      · linarith[hx.2]
+
+/-- If `f'` is the derivative of `f` then  `0 ≤ f' x → (-f)' x ≤ 0`. -/
+theorem deriv_neg_nonpos {f : ℝ → ℝ} {b c : ℝ}
+  (hd₁ : DifferentiableOn ℝ f (Set.Ioo b c))
+  (h₁ : ∀ x ∈ Set.Ioo b c, 0 ≤ deriv f x) (x : ℝ) :
+  x ∈ Set.Ioo b c → deriv (-f) x ≤ 0 := by
+        intro hx
+        show deriv (((fun x => -x) ∘ (fun x => f x))) x ≤ 0
+        rw [deriv.comp]
+        simp
+        apply h₁;tauto
+
+        apply differentiable_neg
+        apply DifferentiableOn.differentiableAt
+        exact hd₁
+        apply Ioo_mem_nhds
+        exact hx.1
+        linarith[hx.2]
+
+/-!
+### The First-Derivative Test
+
+Using the connection beetween monotonicity and derivatives we obtain the familiar
+First-Derivative Test from calculus.
+-/
+
+/-- If `f` is monotone on `(a,b]` and antitone on `[b,c)` then `f` has
+a local maximum at `b`. -/
+lemma isLocalMax_of_mono_anti.{u, v}
```

Comment:
This can go somewhere in `Topology/Order/`

Reply:
Since it builds on both `LocalExtr.lean` and `OrderClosed.lean`, maybe a new file `OrderClosedExtr.lean`?